docs/credstash.md

```diff
@@ -0,0 +1,215 @@
+}
+output "install_snippet" {
+  value = "${module.credstash.install_snippet}"
+}
+output "get_cmd" {
+  value = "${module.credstash.get_cmd}"
+}
+output "put_cmd" {
+  value = "${module.credstash.put_cmd}"
+}
+output "reader_policy_arn" {
+  value = "${module.credstash.reader_policy_arn}"
+}
+output "writer_policy_arn" {
+  value = "${module.credstash.writer_policy_arn}"
+}
+```
+
+## Manual usage
+
+After `credstash` related resources have been setup any user from the list of
+`kms_key_admins` can start using the tool.
+
+Here is a simple example on how to use it on the terminal:
+
+```bash
+$ credstash put my-secret poor-entropy-password
+my-secret has been stored
+$ credstash get my-secret
+poor-entropy-password
+```
+
+Add a new version for existing key (`-a` for auto-increment) and list keys with
+value version numbers:
+
+```
+$ credstash put -a my-secret high-entropy-password
+my-secret has been stored
+$ credstash get my-secret
+high-entropy-password
+$ credstash list
+my-secret      -- version 0000000000000000000
+my-secret      -- version 0000000000000000001
+```
+
+## Automated usage
+
+Considering that roles and users mentioned in above section have administrative
+level of access to KMS Key it would be really bad to use them in EC2 bootstrap
+scripts. In order to narrow down permissions to `credstash-store` DynamoDB table
+and restrict usage patterns to KMS Master
+Key
+[credstash-grant](https://github.com/fpco/fpco-terraform-aws/tree/master/tf-modules/credstash-grant) terrform
+module should be used.
+
+`credstash-grant` module requires an IAM Role(s) ARN and Name as its input,
+which can be created using terraform in a usual way:
+
+```
+resource "aws_iam_instance_profile" "credstash-profile" {
+  name = "credstash-profile"
+  role = "${aws_iam_role.credstash-role.name}"
+}
+resource "aws_iam_role" "credstash-role" {
+  name_prefix  = "credstash-role-"
+  assume_role_policy = <<END_POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "ec2.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+END_POLICY
+}
+```
+
+The purpose of `credstash-grant` module is to automatically assign
+`reader_policy_arn` and/or `writer_policy_arn` to the Role created with above
+code as well as create a KMS reader and/or writer grant(s) for that role. This
+assures that any EC2 instance assuming this role can use credstash only for
+reading and/or writing secrets, but cannot do anything else with neither KMS
+Master Key nor DynamoDB table.
+
+Here is how to give read only permission to that role:
+
+```
+module "credstash-grant" {
+  source            = "github.com/fpco/fpco-terraform-aws/tf-modules/credstash-grant"
+  kms_key_arn       = "${data.terraform_remote_state.credstash.kms_key_arn}"
+  reader_policy_arn = "${data.terraform_remote_state.credstash.reader_policy_arn}"
+  roles_count       = 1
```

Comment:
`roles_count` has to match the length of `roles_arns` and `roles_names`, if it's higher it will result in an error, if less, well, not all roles will be used. I really wish this wan't necessary, but unfortunately if roles are created at the same time this module is used, terraform will not be able to evaluate length of roles and use it as a count. I probably should add a link to the terraform issue about this.